src/backend/c-lisp.py

```diff
@@ -0,0 +1,288 @@
+            instr_list.append(["ret", res_sym])
+            return instr_list
+        else:
+            raise CodegenError(
+                f"Return statement can contain only 1 optional return value: {stmt}"
+            )
+
+    def is_compound_stmt(self, stmt):
+        return isinstance(stmt, list) and isinstance(stmt[0], list)
+
+    def gen_compound_stmt(self, stmt):
+        instr_list = []
+        for s in stmt:
+            instr_list += self.gen_stmt(s)
+        return instr_list
+
+    def is_set_expr(self, expr):
+        return expr[0] == "set"
+
+    def gen_set_expr(self, expr, res_sym):
+        name = expr[1]
+        if not name in self.symbol_types:
+            raise CodegenError(f"Cannot set undeclared variable: {name}")
+
+        instr_list = self.gen_expr(expr[2], res_sym=res_sym)
+        instr_list.append(["set", [name, self.symbol_types[name]], ["id", res_sym]])
+        return instr_list
+
+    def get_literal_type(self, expr):
+        if isinstance(expr, bool):
+            return "bool"
+        elif isinstance(expr, int):
+            return "int"
+        else:
+            return None
+
+    def is_literal_expr(self, expr):
+        return not (self.get_literal_type(expr) is None)
+
+    def gen_literal_expr(self, expr, res_sym):
+        return [["set", [res_sym, self.get_literal_type(expr)], ["const", expr]]]
+
+    def is_call_expr(self, expr):
+        return expr[0] == "call"
+
+    def gen_call_expr(self, expr, res_sym):
+        instr_list = []
+        arg_syms = []
+        for arg in expr[2:]:
+            arg_sym = random_label("tmp_clisp")
+            arg_syms.append(arg_sym)
+            instr_list += self.gen_expr(arg, res_sym=arg_sym)
+        name = expr[1]
+        instr_list.append(
+            ["set", [res_sym, self.function_types[name][0]], ["call", name, *arg_syms]]
+        )
+        return instr_list
+
+    def is_var_expr(self, expr):
+        return isinstance(expr, str)
+
+    def gen_var_expr(self, expr, res_sym):
+        if expr in self.symbol_types:
+            return [["set", [res_sym, self.symbol_types[expr]], ["id", expr]]]
+        else:
+            raise CodegenError(f"Reference to undeclared variable: {expr}")
+
+    def is_binary_expr(self, expr):
+        return expr[0] in self.binary_op_types
+
+    def gen_binary_expr(self, expr, res_sym):
+        if not len(expr) == 3:
+            raise CodegenError(f"Binary operation takes only 2 operands: {expr}")
+
+        instr_list = []
+        in1_sym, in2_sym = [
+            random_label(prefix) for prefix in ("tmp_clisp_in1", "tmp_clisp_in2")
+        ]
+        opcode = expr[0]
+        typ = self.binary_op_types[opcode]
+        return [
+            *self.gen_expr(expr[1], in1_sym),
+            *self.gen_expr(expr[2], in2_sym),
+            ["set", [res_sym, typ], [opcode, in1_sym, in2_sym]],
+        ]
+
+    def gen_expr(self, expr, res_sym=random_label("tmp_clisp")):
```

Comment:
take `res_sym=None` and in the code do something like

```
if res_sym is None:
     res_sym = random_label("tmp_clisp")
```

Reply:
Done